ang/civicase/ActivitiesCalendar.js

```diff
@@ -201,60 +207,62 @@
         return 'invisible';
       }
 
-      if (activities.length === 0 || params.mode !== 'day') {
+      if (!day || params.mode !== 'day') {
         return;
       }
 
-      if (haveAllActivitiesBeenCompleted(activities)) {
-        return 'civicase__activities-calendar__day-status civicase__activities-calendar__day-status--completed';
-      } else if (isAnyActivityOverdue(activities)) {
-        return 'civicase__activities-calendar__day-status civicase__activities-calendar__day-status--overdue';
+      if (day.status === 'completed') {
+        classSuffix = 'completed';
+      } else if (moment(params.date).isSameOrAfter(moment.now())) {
+        classSuffix = 'scheduled';
       } else {
-        return 'civicase__activities-calendar__day-status civicase__activities-calendar__day-status--scheduled';
+        classSuffix = 'overdue';
       }
+
+      return 'civicase__activities-calendar__day-status civicase__activities-calendar__day-status--' + classSuffix;
     }
 
     /**
-     * Returns a list of selected activities for the currently selected date.
+     * Loads the dates with at least an activity with the given status(es)
      *
-     * @return {Array} a list of formatted activities.
+     * @param {*} statusParam
+     * @return {Promise}
      */
-    function getSelectedActivities () {
-      return getActivitiesForDate($scope.selectedDate)
-        .map(function (activity) {
-          return formatActivity(activity, $scope.caseId);
-        });
+    function loadDaysWithActivities (statusParam) {
+      return crmApi('Activity', 'getdayswithactivities', {
+        case_id: $scope.caseId,
+        status_id: statusParam
+      });
     }
 
     /**
-     * Determines if all the given activities have been completed.
+     * Loads the days with at least a completed activity
      *
-     * @param {Array} activities
-     * @return {Boolean}
+     * @return {Promise}
      */
-    function haveAllActivitiesBeenCompleted (activities) {
-      return _.every(activities, function (activity) {
-        return _.includes(CRM.civicase.activityStatusTypes.completed, +activity.status_id);
-      });
+    function loadDaysWithActivitiesCompleted () {
+      return loadDaysWithActivities(CRM.civicase.activityStatusTypes.completed[0])
+        .then(_.curryRight(addDays)('completed'));
     }
 
     /**
-     * Determines if at least one of the given activities is overdue.
+     * Loads the days with at least an incomplete activity
      *
-     * @param {Array} activities
-     * @return {Boolean}
+     * @return {Promise}
      */
-    function isAnyActivityOverdue (activities) {
-      return _.some(activities, function (activity) {
-        return activity.is_overdue;
-      });
+    function loadDaysWithActivitiesIncomplete () {
+      return loadDaysWithActivities({
+        'IN': CRM.civicase.activityStatusTypes.incomplete
+      })
+        .then(_.curryRight(addDays)('incomplete'));
```

Comment:
Sorry, but the logic is a bit hard to understand. Can we break this dow a bit more.
Something like 
```
    function loadDaysWithActivitiesIncomplete () {
         return loadDaysWithActivities({
          'IN': CRM.civicase.activityStatusTypes.incomplete
        })
        .then(function(res){
           // here is the magic spell comment
           _.curryRight(addDays)('incomplete')(res));
      });
```
Something like this.

Reply:
Honestly I don't see why it is harder than your example, we're simply passing a curried function to then, it's pretty standard